backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/JWTgenerator.java

```diff
@@ -0,0 +1,135 @@
+      verifier = new MACVerifier(sharedSecret);
+    } catch (Exception kle) {
+      throw new MolgenisException(kle.getMessage());
+    }
+  }
+
+  /*Token should either have id */
+  public static String createNamedTokenForUser(Database db, String user, String tokenId) {
+    Calendar c = Calendar.getInstance();
+    c.setTime(new Date());
+    c.add(Calendar.YEAR, 10);
+    String token = createNamedTokenForUser(db, user, tokenId, c.getTime());
+    // add token to user settings
+    db.getUser(user).addToken(tokenId);
+    return token;
+  }
+
+  /* Temporary token has an expiration time */
+  public static String createTemporaryToken(Database db, String user) {
+    return createNamedTokenForUser(
+        db, user, "temporary", new Date(new Date().getTime() + 60 * 1000));
+  }
+
+  private static String createNamedTokenForUser(
+      Database db, String user, String tokenId, Date expirationTime) {
+    requireNonNull(user);
+    requireNonNull(tokenId);
+    requireNonNull(expirationTime);
+    if (!db.getActiveUser().equals(ADMIN_USER) && !db.getActiveUser().equals(user)) {
+      throw new MolgenisException("Cannot create token for other users unless admin");
+    }
+    try {
+      if (signer == null) {
+        init();
+      }
+      // Prepare JWT with claims set
+      JWTClaimsSet claimsSet =
+          new JWTClaimsSet.Builder()
+              .subject(user)
+              .jwtID(tokenId)
+              .issueTime(new Date())
+              .expirationTime(expirationTime)
+              .build();
+
+      SignedJWT signedJWT = new SignedJWT(new JWSHeader(JWSAlgorithm.HS256), claimsSet);
+
+      // Apply the HMAC protection
+      signedJWT.sign(signer);
+
+      // Serialize to compact form, produces something like
+      // eyJhbGciOiJIUzI1NiJ9.SGVsbG8sIHdvcmxkIQ.onO9Ihudz3WkiauDO2Uhyuz0Y18UASXlSc1eS0NkWyA
+      return signedJWT.serialize();
+    } catch (JOSEException je) {
+      throw new MolgenisException(je.getMessage());
+    }
+  }
+
+  // parse token
+  public static String getUserFromToken(Database database, String token) {
+    // On the consumer side, parse the JWS and verify its HMAC
+    try {
+      if (signer == null) {
+        init();
+      }
+      SignedJWT signedJWT = SignedJWT.parse(token);
+      Date experationTime = signedJWT.getJWTClaimsSet().getExpirationTime();
+      String tokenId = signedJWT.getJWTClaimsSet().getJWTID();
+      String userName = signedJWT.getJWTClaimsSet().getSubject();
+      if (signedJWT.verify(verifier) && new Date().before(experationTime)) {
+        // verify user is known
+        User user = database.getUser(userName);
+        // if temp token we must verify has experationTime not too far in the future
+        if (user != null && "temporary".equals(tokenId)) {
+          Date thirtyMinutes = new Date(System.currentTimeMillis() + (30 * 60 * 1000));
+          if (experationTime.before(thirtyMinutes)) {
+            return user.getUsername();
+          }
```

Comment:
lots of low level date hacking maybe use (java) lib

Reply:
indeed